server/src/test/java/org/elasticsearch/index/shard/ShardGetServiceTests.java

```diff
@@ -209,4 +212,95 @@ public void testTypelessGetForUpdate() throws IOException {
 
         closeShards(shard);
     }
+
+    public void testGetFromTranslog() throws IOException {
+        Settings settings = indexSettings(Version.CURRENT, 1, 1).build();
+        IndexMetadata metadata = IndexMetadata.builder("test").putMapping("""
+            { "properties": { "foo":  { "type": "text"}}}""").settings(settings).primaryTerm(0, 1).build();
+        IndexShard primary = newShard(new ShardId(metadata.getIndex(), 0), true, "n1", metadata, null);
+        recoverShardFromStore(primary);
+        InternalEngine engine = (InternalEngine) primary.getEngineOrNull();
+
+        // Initially there hasn't been any switches from unsafe to safe maps in the live version map
+        assertEquals(engine.getLastUnsafeSegmentGenerationForGets(), -1);
+        var map = engine.getLiveVersionMap();
+        assertFalse(LiveVersionMapTestUtils.isSafeAccessRequired(map));
+        assertFalse(LiveVersionMapTestUtils.isUnsafe(map));
+
+        // Make the map unsafe by indexing a doc that will be indexed in the append-only mode
+        var indexResult = indexDoc(primary, null, "{\"foo\" : \"baz\"}", XContentType.JSON, "foobar");
+        assertFalse(LiveVersionMapTestUtils.isSafeAccessRequired(map));
+        assertTrue(LiveVersionMapTestUtils.isUnsafe(map));
+
+        // Issue a get that would enforce safe access mode and switches the maps from unsafe to safe
+        var getResult = primary.getService()
+            .getFromTranslog("2", new String[] { "foo" }, true, 1, VersionType.INTERNAL, FetchSourceContext.FETCH_SOURCE, false);
+        assertNull(getResult);
+        var lastUnsafeGeneration = engine.getLastUnsafeSegmentGenerationForGets();
+        assertThat(lastUnsafeGeneration, greaterThan(0L));
+        assertTrue(LiveVersionMapTestUtils.isSafeAccessRequired(map));
+        assertFalse(LiveVersionMapTestUtils.isUnsafe(map));
+
+        // A flush shouldn't change the recorded last unsafe generation for gets
+        PlainActionFuture<Engine.FlushResult> flushFuture = PlainActionFuture.newFuture();
+        engine.flush(true, true, flushFuture);
+        var flushResult = flushFuture.actionGet();
+        assertTrue(flushResult.flushPerformed());
+        assertThat(flushResult.generation(), equalTo(lastUnsafeGeneration));
```

Comment:
I think adding a check that `engine.getLastUnsafeSegmentGenerationForGets()` is equal to `lastUnsafeGeneration` matches the comment a few lines up better. Fine to keep the existing assertion ofc.